src/test/java/com/googleinterns/zoomtube/servlets/LectureServletTest.java

```diff
@@ -0,0 +1,149 @@
+import java.util.List;
+import org.junit.Rule;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import com.ryanharter.auto.value.gson.GenerateTypeAdapter;
+import java.lang.reflect.Type;
+import com.google.gson.reflect.TypeToken;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.mockito.Mock;
+import org.mockito.junit.MockitoJUnit;
+import org.mockito.junit.MockitoRule;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+
+@RunWith(JUnit4.class)
+public final class LectureServletTest {
+  private final LocalServiceTestHelper testServices =
+      new LocalServiceTestHelper(new LocalDatastoreServiceTestConfig());
+  DatastoreService datastoreService;
+  LectureServlet servlet;
+  
+  @Rule public final MockitoRule mockito = MockitoJUnit.rule();
+  @Mock private HttpServletRequest request;
+  @Mock private HttpServletResponse response;
+
+  private static final String LINK_INPUT = "link-input";
+  private static final String TEST_LINK = "https://www.youtube.com/watch?v=wXhTHyIgQ_U";
+
+  @Before
+  public void setUp() throws ServletException {
+    testServices.setUp();
+    datastoreService = DatastoreServiceFactory.getDatastoreService();
+    servlet = new LectureServlet();
+    servlet.init();
+  }
+
+  @After
+  public void tearDown() {
+    testServices.tearDown();
+  }
+
+  @Test
+  public void doPost_urlAlreadyInDatabase_shouldReturnLecture() throws IOException {
+    when(request.getParameter(LINK_INPUT)).thenReturn(TEST_LINK);
+    datastoreService.put(servlet.createLectureEntity(request));
+    
+    servlet.doPost(request, response);
+    
+    assertThat(datastoreService.prepare(new Query("Lecture")).countEntities()).isEqualTo(1);
+    verify(response).sendRedirect("/lecture-view.html?id=1&video-id=wXhTHyIgQ_U");
+  }
+
+  @Test
+  public void doPost_urlNotInDatabase_shouldAddToDatabaseAndReturnRedirect() throws IOException {
+    when(request.getParameter(LINK_INPUT)).thenReturn(TEST_LINK);
+   
+    servlet.doPost(request, response);
+    
+    assertThat(datastoreService.prepare(new Query("Lecture")).countEntities()).isEqualTo(1);
+    verify(response).sendRedirect("/lecture-view.html?id=1&video-id=wXhTHyIgQ_U");
+  }
+
+  @Test
+  public void doGet_emptyDatabase_shouldReturnNoLecture() throws IOException {
+    StringWriter content = new StringWriter();
+    PrintWriter writer = new PrintWriter(content);
+    when(response.getWriter()).thenReturn(writer);
+
+    servlet.doGet(request, response);
+    
+    String json = content.toString();
+    assertThat(json).startsWith("[]");
+  }
+
+  @Test
+  public void doGet_emptyDatabase_shouldReturnOneLecture() throws IOException {
+    when(request.getParameter(LINK_INPUT)).thenReturn(TEST_LINK);
```

Comment:
What does this param do in a GET request?

Reply:
It is used to add an entity to the database

Reply:
I thought GET only ever responds with the list of available lectures.  Is it also used to submit a new one?

Reply:
NVM request is used to create entity in .put